### PR TITLE
[#12] [Backend] As a user, I can see the list of surveys

### PR DESCRIPTION
**Pull request by Tuubz**
Closes https://github.com/nimblehq/ic-flutter-taher-toby/issues/12

## What happened 👀

- Remove old classes (`credential_repository.dart`,  `api_service.dart`, ...)
- Add `survey_service.dart` for fetching the survey list (with additional parameters: `pageNumber` & `pageSize`)
- Add `survey_repository.dart` for calling the service (later we will do caching here)
- Add `get_surveys_use_case.dart` for calling the repository (for testing, I added it to the "Next" button)
- Add `Authorization` header in `app_interceptor.dart` for testing
- Add tests for the repository and use-case (also had to set-up mock classes and files)

## Insight 📝

This is how the back-end flow works:

1. Call API, which returns a JSON string
2. Flatten (simplify) the JSON string, by decoding it with [japx](https://pub.dev/packages/japx)
3. Create a response object from the JSON string, by converting it with [json_annotation](https://pub.dev/packages/json_annotation)
4. Create a model object from the response object, so we can handle the default values in case it's `null`

**Note**: Because the authentication feature hasn't been implemented yet, I generate the token from Postman 🙏 

## Proof Of Work 📹

```
I/flutter (10992): *** Request ***
I/flutter (10992): uri: https://survey-api.nimblehq.co/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=5
I/flutter (10992): method: GET
I/flutter (10992): responseType: ResponseType.json
I/flutter (10992): followRedirects: true
I/flutter (10992): connectTimeout: 3000
I/flutter (10992): sendTimeout: 0
I/flutter (10992): receiveTimeout: 5000
I/flutter (10992): receiveDataWhenStatusError: true
I/flutter (10992): extra: {}
I/flutter (10992): headers:
I/flutter (10992):  Authorization: Bearer iUqrxx2OZmnlRhjpZzQiG6Bhoci7OnA6L0AMIa7giqc
I/flutter (10992): data:
I/flutter (10992): {}

I/flutter (10992): *** Response ***
I/flutter (10992): uri: https://survey-api.nimblehq.co/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=5
I/flutter (10992): statusCode: 200
I/flutter (10992): headers:
I/flutter (10992):  connection: keep-alive
I/flutter (10992):  cache-control: max-age=0, private, must-revalidate
I/flutter (10992):  transfer-encoding: chunked
I/flutter (10992):  date: Thu, 30 Mar 2023 05:50:20 GMT
I/flutter (10992):  vary: Accept-Encoding, Origin
I/flutter (10992):  content-encoding: gzip
I/flutter (10992):  strict-transport-security: max-age=31536000; includeSubDomains
I/flutter (10992):  referrer-policy: strict-origin-when-cross-origin
I/flutter (10992):  report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=mWn9nqWNtsRznd39WdQzv1Z6YRVk8hbbSq0IZAPdHqCbVdPPvao%2FfCv%2FF7lXYpsaw5I1aZQQUibAv6Z%2BdBBLk8q1EUvP073YVFeEFuSri988llXpIkVMomPQwY3s5ZeyVxGf%2Fdlbjq4i"}],"group":"cf-nel","max_age":604800}
I/flutter (10992):  cf-cache-status: DYNAMIC
I/flutter (10992):  x-permitted-cross-domain-policies: none
I/flutter (10992):  content-type: application/json; charset=utf-8
I/flutter (10992):  x-xss-protection: 1; mode=block
I/flutter (10992):  server: cloudflare
I/flutter (10992):  x-request-id: c2787bf7-5b27-4073-b146-8ddc00273d50
I/flutter (10992):  alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
I/flutter (10992):  nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
I/flutter (10992):  x-download-options: noopen
I/flutter (10992):  cf-ray: 7afe130c8fab45cb-BKK
I/flutter (10992):  x-runtime: 0.046778
I/flutter (10992):  etag: W/"ca4ce687b7efc365d85536f7d1c110ff"
I/flutter (10992):  via: 1.1 vegur
I/flutter (10992):  x-frame-options: SAMEORIGIN
I/flutter (10992):  x-content-type-options: nosniff
I/flutter (10992): Response Text:
I/flutter (10992): {"data":[{"id":"d5de6a8f8f5f1cfe51bc","type":"survey_simple","attributes":{"title":"Scarlett Bangkok","description":"We'd love ot hear from you!","thank_email_above_threshold":"<span style=\"font-family:arial,helvetica,sans-serif\"><span style=\"font-size:14px\">Dear {name},<br /><br />Thank you for visiting Scarlett Wine Bar &amp; Restaurant at Pullman Bangkok Hotel G &nbsp;and for taking the time to complete our guest feedback survey!<br /><br />Your feedback is very important to us and each survey is read individually by the management and owners shortly after it is sent. We discuss comments and suggestions at our daily meetings and use them to constantly improve our services.<br /><br />We would very much appreciate it if you could take a few more moments and review us on TripAdvisor regarding your recent visit. By <a href=\"https://www.tripadvisor.com/Restaurant_Review-g293916-d2629404-Reviews-Scarlett_Wine_Bar_Restaurant-Bangkok.html\">clicking here</a> you will be directed to our page.&nbsp;<br /><br 
```

### CODE REVIEW

File: lib/api/repository/survey_repository.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import 'package:flutter_survey/api/survey_service.dart';
+import 'package:flutter_survey/api/exception/network_exceptions.dart';
+import 'package:flutter_survey/model/surveys_model.dart';
+import 'package:injectable/injectable.dart';
+
+abstract class SurveyRepository {
+  Future<SurveysModel> getSurveys(
+      {required int pageNumber, required int pageSize});
```
Comment (Shayokh144) [inline lib/api/repository/survey_repository.dart:8]:
```suggestion
      {required int pageNumber, required int pageSize}
      );
```

File: lib/model/surveys_model.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  @override
+  List<Object?> get props => [surveys];
+
+  factory SurveysModel.fromResponse(SurveysResponse response) {
+    final surveyList = response.surveys ?? List.empty();
+    return SurveysModel(
+        surveys: surveyList
+            .map((survey) => SurveyModel.fromResponse(survey))
+            .toList());
```
Comment (Shayokh144) [inline lib/model/surveys_model.dart:20]:
```suggestion
            .toList()
            );
```

Comment (Tuubz) [reply lib/model/surveys_model.dart:20]: @Shayokh144 Hmmmm, strange

Is this what VStudio is recommending? 💭 

I'm using the default format for Dart in Android Studio

**Review by Shayokh144 — APPROVED** [review]
looks good to me

File: lib/api/repository/survey_repository.dart (hunk 1 above)
Comment (luongvo) [reply lib/api/repository/survey_repository.dart:8]: better to format it like this, by adding 1 more `,` after `pageSize` param + reformatting

```
  Future<SurveysModel> getSurveys({
    required int pageNumber,
    required int pageSize,
  });
```

File: lib/api/response/survey_response.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'survey_response.g.dart';
+
+@JsonSerializable()
+class SurveyResponse {
+  String? id;
+  String? title;
```
Comment (luongvo) [inline lib/api/response/survey_response.dart:8]: We will need to 

- switch the `background`

image

- and show the `description`

Please check again the necessary fields for the following tasks.

File: lib/di/interceptor/app_interceptor.dart [hunk 4]
```diff
@@ -2,22 +2,24 @@ import 'dart:io';
...
   Future onRequest(
       RequestOptions options, RequestInterceptorHandler handler) async {
-    if (_requireAuthenticate) {
-      // TODO header authorization here
-      // options.headers
-      //     .putIfAbsent(HEADER_AUTHORIZATION, () => "");
+    if (_requireAuthentication) {
+      // TODO: Integrate log-in https://github.com/nimblehq/ic-flutter-taher-toby/issues/10
+      options.headers.putIfAbsent(
+          _authorizationHeader, () => "Bearer add your token here");
```
Comment (luongvo) [inline lib/di/interceptor/app_interceptor.dart:22]: I'm curious how can you verify the get surveys API with the Authentication required without doing the Authentication API integration first. It's a bit in the wrong order of implementation, right?

File: lib/di/interceptor/app_interceptor.dart [hunk 5]
```diff
@@ -27,7 +29,7 @@ class AppInterceptor extends Interceptor {
     final statusCode = err.response?.statusCode;
     if ((statusCode == HttpStatus.forbidden ||
             statusCode == HttpStatus.unauthorized) &&
-        _requireAuthenticate) {
+        _requireAuthentication) {
```
Comment (luongvo) [inline lib/di/interceptor/app_interceptor.dart:32]: 👍

File: lib/model/survey_model.dart [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  final String id;
+  final String title;
+
+  const SurveyModel({required this.id, required this.title});
+
+  @override
+  List<Object?> get props => [id, title];
+
+  factory SurveyModel.fromResponse(SurveyResponse response) {
+    return SurveyModel(id: response.id ?? "", title: response.title ?? "");
```
Comment (luongvo) [inline lib/model/survey_model.dart:14]: 1 more field so this should be re-formatted with each field in each line of code 🙏

File: lib/model/surveys_model.dart [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import 'package:equatable/equatable.dart';
+import 'package:flutter_survey/api/response/surveys_response.dart';
+import 'package:flutter_survey/model/survey_model.dart';
+
+class SurveysModel extends Equatable {
+  final List<SurveyModel> surveys;
+
+  const SurveysModel({
+    required this.surveys,
+  });
```
Comment (luongvo) [inline lib/model/surveys_model.dart:10]: Do we need to have this model? Why don't we use `List<SurveyModel>` as a type in other places to simplify this?

File: lib/model/surveys_model.dart (hunk 2 above)
Comment (luongvo) [inline lib/model/surveys_model.dart:20]: This simple mapping logic could be moved to the Repository. It's not so complicated to have an extension, or inside a separate `SurveysModel` class 🤷‍♂️

File: lib/usecases/get_surveys_use_case.dart [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+@Injectable()
+class GetSurveysUseCase extends UseCase<List<SurveyModel>, GetSurveysInput> {
+  final SurveyRepository _surveyRepository;
+
+  const GetSurveysUseCase(this._surveyRepository);
+
+  @override
+  Future<Result<List<SurveyModel>>> call(GetSurveysInput input) {
```
Comment (luongvo) [inline lib/usecases/get_surveys_use_case.dart:24]: Here, why don't we use `SurveysModel` here? Should we use `List<SurveyModel>` for all and get rid of `SurveysModel`? 😕

**Review by luongvo — COMMENTED** [review]
The rest lgtm

File: lib/api/response/survey_response.dart (hunk 3 above)
Comment (Tuubz) [reply lib/api/response/survey_response.dart:8]: Aah yes, correct. 🙏 

I should have added them in this PR, please allow me to update these changes in the [integration PR](https://github.com/nimblehq/ic-flutter-taher-toby/pull/56) 🙇

File: lib/di/interceptor/app_interceptor.dart (hunk 4 above)
Comment (Tuubz) [reply lib/di/interceptor/app_interceptor.dart:22]: Yes, you're right. I should have worked on the authentication first

But it's OK, I authenticate with help of Postman to get the token first, and then paste it in here when testing 🙏 

cc @Shayokh144

Comment (luongvo) [reply lib/di/interceptor/app_interceptor.dart:22]: @Tuubz I don't see that comment in the PR so I need to ask you about this point. That's okay, but please let the reviewers know what's not in the normal process, the Insight section is made for that purpose too 🙏

Comment (Tuubz) [reply lib/di/interceptor/app_interceptor.dart:22]: Yes, thank you for the reminder.

I have updated the Insights of this PR 🙏

File: lib/model/surveys_model.dart (hunk 7 above)
Comment (Tuubz) [reply lib/model/surveys_model.dart]: You're right, it was a bit too redundant, I have removed it 🙏

File: lib/usecases/get_surveys_use_case.dart (hunk 8 above)
Comment (Tuubz) [reply lib/usecases/get_surveys_use_case.dart:24]: @luongvo Updated, I have removed `SurveysModel` 🚀

**Review by luongvo — APPROVED** [review]

**Pull request merged**